pkg/autoscaler/autoscaler.go

```diff
@@ -79,51 +84,62 @@ type totalAggregation struct {
 func (agg *totalAggregation) aggregate(stat Stat) {
 	current, exists := agg.perPodAggregations[stat.PodName]
 	if !exists {
-		current = &perPodAggregation{}
+		current = &perPodAggregation{window: agg.window}
 		agg.perPodAggregations[stat.PodName] = current
 	}
-	current.aggregate(stat.AverageConcurrentRequests)
-	if stat.PodName == ActivatorPodName {
-		agg.containsActivatorMetrics = true
+	if stat.LameDuck {
+		current.lameduck(stat.Time)
+	} else {
+		current.aggregate(stat.AverageConcurrentRequests)
+		if stat.PodName == ActivatorPodName {
+			agg.containsActivatorMetrics = true
+		}
+		agg.probeCount += 1
 	}
-	agg.probeCount++
 }
 
 // The number of pods that are observable via stats
 // Substracts the activator pod if its not the only pod reporting stats
-func (agg *totalAggregation) observedPods() int {
-	observedPods := len(agg.perPodAggregations)
+func (agg *totalAggregation) observedPods(now time.Time) float64 {
+	podCount := float64(0.0)
+	for _, pod := range agg.perPodAggregations {
+		podCount += pod.usageRatio(now)
+	}
 	if agg.containsActivatorMetrics {
-		if observedPods <= 1 {
-			return 1
+		// Discount the activator in the pod count
+		if podCount > 1.0 {
+			return podCount - 1.0
 		}
-		return observedPods - 1
 	}
-	return observedPods
+	return podCount
 }
 
 // The observed concurrency per pod (sum of all average concurrencies
 // distributed over the observed pods)
 // Ignores activator sent metrics if its not the only pod reporting stats
-func (agg *totalAggregation) observedConcurrencyPerPod() float64 {
+func (agg *totalAggregation) observedConcurrencyPerPod(now time.Time) float64 {
 	accumulatedConcurrency := float64(0)
-	podsInCalculation := 0
-
-	observedPods := len(agg.perPodAggregations)
-
+	activatorConcurrency := float64(0)
+	observedPods := agg.observedPods(now)
 	for podName, perPod := range agg.perPodAggregations {
-		if podName != ActivatorPodName || observedPods == 1 {
-			accumulatedConcurrency += perPod.calculateAverage()
-			podsInCalculation++
+		if podName == ActivatorPodName {
+			activatorConcurrency += perPod.calculateAverage(now)
+		} else {
+			accumulatedConcurrency += perPod.calculateAverage(now)
 		}
 	}
-	return accumulatedConcurrency / float64(podsInCalculation)
+	if accumulatedConcurrency == 0.0 {
+		return activatorConcurrency / observedPods
+	}
+	return accumulatedConcurrency / observedPods
```

Comment:
Duh, that's a much simpler piece of logic than what I came up with 👍

Reply:
I had to look at it for a while.

I wonder if it would be better to report the metrics coming from the Activator as just a "backlog", only reporting those requests which haven't started proxying.  Then we can take the backlog into account explicitly in the Autoscaler.

The "if pod == activator" logic seems strange in counting pods.  And the sudden switch from activator metrics to pod metrics may still have some unforseen consequences.

Reply:
IIRC my rationale was that there is no such backlog in the activator. I might need to check again but the behavior was that all requests are retrying in parallel and not sequentially (which makes sense? Once the revision is ready to serve, we should fire everything at it that we can).

Effectively that means that there is no such backlog really?

We should surely revisit this logic though, especially if we start refunctioning the activator in something like the KBuffer.

Reply:
The backlog for a given revision is the length of the list of waiting requests: https://github.com/knative/serving/blob/ad6a0b95aada927b076abb25c584fadb5300190e/pkg/activator/dedupe.go#L44

Once a request is proxied to a pod, this backlog is drained and the load is represented as concurrency on the pod.

I think this would be a better model for the KBuffer because then we would use the metrics flowing back from the Autoscaler about capacity to rate limiting the draining of the backlog.

Reply:
Of course if we use it this way, we should rename `dedupe.go` to `queue.go` or something like that.